### ISSUE TITLE
Remove callWithLogger usage in reactors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6789]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6789|
|Type|task|
|Created|2013-10-16 06:25:22Z|
|Branch|https://github.com/twisted/twisted/tree/no-call-with-logger-6789-2|

Some people have been complaining about the performance impact of callWithLogger when used in the reactor. The whole context idea is fairly clunky, so removing it usage seems like a good idea. That said, it's a backward incompatible change, so we need to agree before moving. This ticket should contain the relevant discussion.

<details><summary>Searchable metadata</summary>

```
trac-id__6789 6789
type__task task
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_no_call_with_logger_6789_2 branches/no-call-with-logger-6789-2
branch_author__hawkowl__therve hawkowl, therve
status__new new
resolution__None None
component__core core
keywords__ 
time__1381904722000000 1381904722000000
changetime__1458322289801078 1458322289801078
version__None None
owner__therve therve
cc__wsanchez cc__glyph cc__exarkun cc__twm@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [40363]) Branching to 'no-call-with-logger-6789'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

I would also note that:
 * callWithLogger is also used in Conch.
 * It should be deprecated, alongside callWithContext.
 * The whole twisted.python.context module should be deprecated.

Those things should happen in other branches. We should also talk about the default log format: if "system" is not passed anymore, it's worth checking whether we want to remove the corresponding part of the format.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Alright, buildbot is mostly happy, feedback welcome!

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Alex commented|
|-|-|

Reviewed the patch, and I believe it all looks correct.  There's also a sample traceback in `doc/mail/tutorial/smtpclient/smtpclient.xhtml` that you probably want to remove the `callWithLogger` frames from. Other than that I think this looks good to me. Not going to unmark-reviewed as I'm not totally comfortable with that yet.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Oh, I must add that I run some Twisted benchmarks. In particular, tcp_throughput got about 25% faster compared to trunk.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_6"></a>|@dreid commented|
|-|-|

> We should also talk about the default log format: if "system" is not passed anymore, it's worth checking whether we want to remove the corresponding part of the format.

I wouldn't change the current format at all.  With this change the default will be more often '-' but changing the default would break expectations for people who actively set the system (as I have done and as I have suggested people on #twisted do.)

I am concerned about the loss of some of the structured data that has been mis-exposed in the system portion of the format, such as the session number and peer address as set by https://github.com/twisted/twisted/blob/trunk/twisted/internet/tcp.py#L779.

I believe there are probably tools that exist in the wild in places we can't possibly know about that make use of this information.  However I am completely comfortable with just deciding that the human readable logs are _not_ a surface where we have to maintain compatibility as they have never been _designed_ to be machine readable.  

The system field, which is what is ultimately set by the use of callWithLogger, is also wildly incorrect whenever system is not set explicitly so I'd say this change makes the system field less-incorrect more of the time.

> In particular, tcp_throughput got about 25% faster compared to trunk.

Was this on CPython or PyPy?


I'm a +1 on this change, the only technical comment I have is that we are very inconsistent about when we prefer a local to an attribute lookup among our reactors and it'd be nice if we fixed that because _drdw isn't a very good name for a thing in general and that can always be deferred to another ticket.


But we need to talk more about how this fits in with our compatibility policy before I'm comfortable saying that we should merge it.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_7"></a>|@radix commented|
|-|-|

Are there really no other compatibility "surfaces" other than the format of the log file? Will absolutely no other code break with this change? I guess you could have an observer that expects system and now won't be receiving it. anything else?

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Alex commented|
|-|-|

No, observers will still receive a ``'system'`` key, it will just always have ``'-'`` for a value, since a default is set.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> No, observers will still receive a 'system' key, it will just always have '-' for a value, since a default is set.

Which is no more an incompatible change than, say, changing `dataReceived` so it is always called with `b"foo"` instead of the data read from the socket. :)

I think it's futile to try to invent an explanation for how this isn't an incompatible change.

Instead, the way to justify this change is to make it an explicit exception to the [compatibility policy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy).  This can probably be done on the grounds that it's a very obscure API which little or no known real-world usage outside of Twisted itself.  Perhaps we can increase our confidence that this is the case by soliciting community feedback regarding its removal.  Simultaneously, we can alert the community about this upcoming change and give anyone with code that does rely on it a head-start on removing that reliance (supposing that they're not interesting in arguing against the removal).

For what it's worth, I'm curious about how the new logging system [ticket:6750 currently under development] elsewhere will address use-cases like this one.  I don't think anyone is suggesting this removal makes sense because the information of where a log event originates isn't *useful* - merely that the current implementation of this feature is inadequate.  It would be nice to have some idea how a future implementation might be made adequate.  Since the new logging library is still incomplete, perhaps it is too early to be able to answer this question (in which case I wonder if it is too early to remove this code - but not strongly enough to suggest one task must be blocked on the other).

Separately, I notice that the branch doesn't come with a `.removal` [news fragment](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).  This seems particularly important if the change is accepted as a policy violation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_9):

> For what it's worth, I'm curious about how the new logging system [ticket:6750 currently under development] elsewhere will address use-cases like this one.  I don't think anyone is suggesting this removal makes sense because the information of where a log event originates isn't *useful* - merely that the current implementation of this feature is inadequate.  It would be nice to have some idea how a future implementation might be made adequate.  Since the new logging library is still incomplete, perhaps it is too early to be able to answer this question (in which case I wonder if it is too early to remove this code - but not strongly enough to suggest one task must be blocked on the other).

The current status of that branch is that if you're using the old style of logging, it's the same, but if you use the new style, the 'system' key is just always replaced with the namespace of the logger doing the logging.  I believe that Wilfredo's contention (which I am somewhat sympathetic to but maybe don't entirely agree with) is that the 'system' key is almost always useless noise, and in the cases where it isn't, it should be communicated some other way (like having an instance-level bound Logger that remembers your current HTTP request or IRC user or whatever).

If people still feel like the current usage / value of the 'system' key is useful in some circumstances, we could possibly do it more efficiently, which is to say, making it slightly inefficient to retrieve it the first time an iteration logs the message by doing stack-walking, rather than making it inefficient to simply iterate the reactor _in case_ anyone might want to log a message.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_9):

> Instead, the way to justify this change is to make it an explicit exception to the [compatibility policy](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy).  This can probably be done on the grounds that it's a very obscure API which little or no known real-world usage outside of Twisted itself.  Perhaps we can increase our confidence that this is the case by soliciting community feedback regarding its removal.  Simultaneously, we can alert the community about this upcoming change and give anyone with code that does rely on it a head-start on removing that reliance (supposing that they're not interesting in arguing against the removal).

Since we've both told people on several occasions that we can have exceptions to that policy if there is sufficient discussion / consensus about it not breaking any real code, but the policy itself does not explain how such an exception might be made or what would be considered sufficient discussion, I've amended that policy page to include [a new section on some acceptable parameters for discussion about exceptions](https://twistedmatrix.com/trac/wiki/CompatibilityPolicy#ProcedureforExceptionstothisPolicy), based on my recollections of prior consensus.

If anyone takes issue with the mechanisms I've proposed there feel free to edit it, the yardsticks chosen there were somewhat arbitrary.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez commented|
|-|-|

I should clarify glyph’s response to exarkun a bit.  The new logging module’s (that is, the one I’m working on in [#6750](https://github.com/twisted/twisted/issues/6750)) `FileLogObserver`, which is where "the default log format” is defined, derives the “system” from the namespace and level of the emitter, and has no knowledge of `callWithContext`.

However, if you explicitly set the `log_system` key (or `system` if using the old API), then that is used instead.

As Glyph mentioned, I have a hard time finding the existing system key as populated by `callWithContext`, and believe that there are far more useful ways for a developer to specify this sort of information.

In support of this ticket, I certainly believe reactor should have no knowledge of the logging system at all, and it certainly shouldn’t be doing extra (and apparently expensive) work for the potential (and in practice unlikely) benefit of specific higher-level code like logging.

The new logging module is actually pretty close to complete now, I think.  What’s left is documentation (I’m still working on a HOWTO) and making sure it passes on all builders, so the timing may be more in line with this ticket than not (assuming it passes review).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez set owner to @therve|
|-|-|

No builds in buildbot yet, but `twisted/internet/test/test_udp.py` has svn conflict turds in it.

Otherwise, a read of the code looks pretty straightforward.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

(In [47007]) Branching to no-call-with-logger-6789-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

Unfortunately this can't land in the present state.

When this patch is merged forward, running something like twistd manhole gives the following:

```
2016-03-17 19:52:52+0800 [-] Log opened.
2016-03-17 19:52:52+0800 [-] twistd 16.0.0 (/Users/red/venvs/commands/bin/python 2.7.10) starting up.
2016-03-17 19:52:52+0800 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2016-03-17 19:52:52+0800 [-] ConchFactory starting on 8000
2016-03-17 19:52:52+0800 [-] Starting factory <twisted.conch.manhole_ssh.ConchFactory instance at 0x10c9ead40>
2016-03-17 19:52:59+0800 [-] disabling non-fixed-group key exchange algorithms because we cannot find moduli file
2016-03-17 19:52:59+0800 [-] Disconnecting with error, code 2
	reason: bad packet length 1195725856
2016-03-17 19:52:59+0800 [-] connection lost
```

Compare this to trunk:

```
2016-03-17 19:53:18+0800 [-] Log opened.
2016-03-17 19:53:18+0800 [-] twistd 16.0.0 (/Users/red/venvs/commands/bin/python 2.7.10) starting up.
2016-03-17 19:53:18+0800 [-] reactor class: twisted.internet.selectreactor.SelectReactor.
2016-03-17 19:53:18+0800 [-] ConchFactory starting on 8000
2016-03-17 19:53:18+0800 [-] Starting factory <twisted.conch.manhole_ssh.ConchFactory instance at 0x105ae3ef0>
2016-03-17 19:53:22+0800 [twisted.conch.manhole_ssh.ConchFactory] disabling non-fixed-group key exchange algorithms because we cannot find moduli file
2016-03-17 19:53:22+0800 [SSHServerTransport,0,127.0.0.1] Disconnecting with error, code 2
	reason: bad packet length 1195725856
2016-03-17 19:53:22+0800 [SSHServerTransport,0,127.0.0.1] connection lost
```

As you can see, some weirdness occurs because of the lack of log context, even when the logger is actually a new-style logger (the twisted.conch.manhole_ssh.ConchFactory log line is because protocol.Factory uses the new logger).

I think maybe twistd's logger being old-style is contributing to this?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_17"></a>|@wsanchez commented|
|-|-|

Yeah, most of Twisted uses the old emitter API and registers observers with the old API, so even if you are using the new API, the old machinery is still in play.

I'd be interested in working with someone on porting twistd (and trial) to the new API at PyCon if nobody takes it on before then.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl commented|
|-|-|

twistd porting to new logger up for review: [#8235](https://github.com/twisted/twisted/issues/8235). I don't think trial uses the logger?

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_19"></a>|@wsanchez commented|
|-|-|

`trial` does not.  I started and got stuck on an existing test that I don't fully understand; see [#7863](https://github.com/twisted/twisted/issues/7863).
